staging/src/k8s.io/client-go/util/workqueue/queue_test.go

```diff
@@ -159,3 +159,106 @@ func TestReinsert(t *testing.T) {
 		t.Errorf("Expected queue to be empty. Has %v items", a)
 	}
 }
+
+func TestQueueDrainageUsingShutDownWithDrain(t *testing.T) {
+
+	q := workqueue.New()
+
+	q.Add("foo")
+	q.Add("bar")
+
+	firstItem, _ := q.Get()
+	secondItem, _ := q.Get()
+
+	finishedWG := sync.WaitGroup{}
+	finishedWG.Add(1)
+	go func() {
+		defer finishedWG.Done()
+		q.ShutDownWithDrain()
+	}()
+
+	// Add as many items as possible until the queue is marked "ShuttingDown"
```

Comment:
State the purpose of this? Is it to exercise a race condition? If so it might be good to have it run this test many times.

Reply:
The purpose of this was to ensure a sequence of events where `ShutDownWithDrain` is called before we start marking items as done - thus "simulating" that we drain the queue / wait for all items to finish processing. I realized that the way of doing this was a bit contrived given my initial goal and I have now re-written it, removing `deltaItems`: which should make the purpose clearer.